<a href="https://colab.research.google.com/github/Lisepetrova/Final_PythonUrban/blob/main/%D0%9F%D0%B5%D1%82%D1%80%D0%BE%D0%B2%D0%B0_%D0%98%D1%82%D0%BE%D0%B3%D0%BE%D0%B2%D0%BE%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Итоговый проект. Улицы. Объекты культурного наследия.

In [82]:

!pip install folium -U
!pip install geopandas mapclassify
!pip install osmnx

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [83]:
import pandas as pd
import geopandas as gpd
import osmnx as ox

In [84]:
pd.options.plotting.backend = "plotly"

In [85]:
TILES = "cartodbpositron"


DISTRICT_NAME = 'Пушкинский район'
TERRITORY_NAME = 'Пушкинский район, Санкт-Петербург'
STREETS_FILE_URL = "https://drive.google.com/file/d/1bUT1E-QSbG1vpSNM2dOG2-LEVXSrPdo3/view?usp=sharing"
KGIOP_FILE_URL = "https://raw.githubusercontent.com/aeksei/python-urban/main/kgiop_objects.geojson"

In [86]:
def get_google_drive_download_url(url: str) -> str:
    """
    Функция возвращает ссылку для скачивания с google drive.
     Например: "https://drive.google.com/file/d/1MWGOuqV76e0ubQOg8Ke0KTU3yGpkprHp/view?usp=sharing" ->
     "https://drive.google.com/uc?export=download&id=1MWGOuqV76e0ubQOg8Ke0KTU3yGpkprHp"
    """
    drive_id = url.split("/")[5]
    return f"https://drive.google.com/uc?export=download&id={drive_id}"

## Территория

### Загрузка территории из OSM (Extract)

In [87]:
territory = ox.geocode_to_gdf(TERRITORY_NAME)
territory.explore(tiles=TILES)

In [88]:
territory = ox.geocode_to_gdf(TERRITORY_NAME)  # подгружаем с OSM и получаем GeoDataFrame
territory

geometry  bbox_north  bbox_south  \
0  POLYGON ((30.21822 59.67307, 30.22257 59.67064...   59.829374   59.633783   

   bbox_east  bbox_west   place_id  osm_type  osm_id        lat        lon  \
0  30.576345  30.218223  298402523  relation  338635  59.731838  30.424492   

                                        display_name     class  \
0  Пушкинский район, Pushkin, Фёдоровское городск...  boundary   

             type  importance  
0  administrative    0.560862

### Обработка данных с территорией (Transform)

In [89]:
drop_columns = [
    "bbox_north",
    "bbox_south",
    "bbox_east",
    "bbox_west",
    "place_id",
    "osm_type",
    "osm_id",
    "lat",
    "lon",
    "display_name",
    "class",
    "type",
    "importance"
]
territory.drop(columns=drop_columns, inplace=True)

In [90]:
territory['Название территории'] = ['Пушкинский район']
territory

geometry Название территории
0  POLYGON ((30.21822 59.67307, 30.22257 59.67064...    Пушкинский район

In [91]:
territory_streets = territory.dissolve(by = "Название территории")
territory_streets

geometry
Название территории                                                   
Пушкинский район     POLYGON ((30.21822 59.67307, 30.22257 59.67064...

## Улицы

### Загрузка файла с улицами из google disk (Extract)

In [92]:
STREETS_FILE_URL = "https://drive.google.com/file/d/1bUT1E-QSbG1vpSNM2dOG2-LEVXSrPdo3/view?usp=sharing"
url = get_google_drive_download_url(STREETS_FILE_URL)
gdf_streets = gpd.read_file(url, mask=territory)  
gdf_streets

type         id                                               tags  \
0     way    4414959  {'highway': 'residential', 'lanes': '2', 'lit'...   
1     way    4414960  {'highway': 'residential', 'lanes': '2', 'lit'...   
2     way    4414988  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
3     way    4456239  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
4     way    4460441  {'destination': 'Россия;Москва', 'destination:...   
...   ...        ...                                                ...   
8327  way  982201974                             {'highway': 'service'}   
8328  way  982201975                             {'highway': 'service'}   
8329  way  982201976        {'access': 'private', 'highway': 'service'}   
8330  way  982309906  {'highway': 'service', 'psv': 'yes', 'surface'...   
8331  way  982465716  {'foot': 'no', 'highway': 'secondary', 'lanes'...   

                                               geometry  
0     LINESTRING (3386560.825 8339904.673, 3386543.1...  
1     LINESTRING (3386296.074 8339493.502, 3386326.1...  
2     LINESTRING (3385061.151 8336249.836, 3385080.0...  
3     LINESTRING (3383520.646 8339537.624, 3383522.2...  
4     LINESTRING (3380805.073 8358710.316, 3380674.5...  
...                                                 ...  
8327  LINESTRING (3387389.855 8336056.523, 3387395.9...  
8328  LINESTRING (3399572.860 8344030.830, 3399577.6...  
8329  LINESTRING (3399513.438 8343712.428, 3399614.3...  
8330  LINESTRING (3386841.039 8339837.185, 3386808.8...  
8331  LINESTRING (3383406.499 8342087.601, 3383409.9...  

[8332 rows x 4 columns]

### Обработка данных с улицами (Transform)

In [93]:
tags = pd.json_normalize(gdf_streets["tags"])
tags

highway lanes  lit  maxspeed                       name oneway  \
0     residential     2  yes  RU:urban   бульвар Алексея Толстого    yes   
1     residential     2  yes  RU:urban          Ахматовская улица    NaN   
2       secondary     2  yes  RU:urban              Садовая улица    NaN   
3       secondary     1  yes  RU:urban  площадь Египетские Ворота    yes   
4           trunk     1  yes  RU:urban                        NaN    yes   
...           ...   ...  ...       ...                        ...    ...   
8327      service   NaN  NaN       NaN                        NaN    NaN   
8328      service   NaN  NaN       NaN                        NaN    NaN   
8329      service   NaN  NaN       NaN                        NaN    NaN   
8330      service   NaN  NaN       NaN                        NaN    NaN   
8331    secondary     3  yes  RU:urban        Петербургское шоссе    yes   

      surface width foot parking:lane:both  ... motorway traffic_sign  \
0     asphalt    10  NaN               NaN  ...      NaN          NaN   
1     asphalt   NaN  NaN               NaN  ...      NaN          NaN   
2     asphalt   NaN   no       no_stopping  ...      NaN          NaN   
3     asphalt   NaN   no               NaN  ...      NaN          NaN   
4     asphalt   NaN  NaN               NaN  ...      NaN          NaN   
...       ...   ...  ...               ...  ...      ...          ...   
8327      NaN   NaN  NaN               NaN  ...      NaN          NaN   
8328      NaN   NaN  NaN               NaN  ...      NaN          NaN   
8329      NaN   NaN  NaN               NaN  ...      NaN          NaN   
8330  asphalt    14  NaN               NaN  ...      NaN          NaN   
8331  asphalt    12   no               NaN  ...      NaN          NaN   

     lanes:forward turn:backward waterway proposed:oneway turn footway  \
0              NaN           NaN      NaN             NaN  NaN     NaN   
1              NaN           NaN      NaN             NaN  NaN     NaN   
2              NaN           NaN      NaN             NaN  NaN     NaN   
3              NaN           NaN      NaN             NaN  NaN     NaN   
4              NaN           NaN      NaN             NaN  NaN     NaN   
...            ...           ...      ...             ...  ...     ...   
8327           NaN           NaN      NaN             NaN  NaN     NaN   
8328           NaN           NaN      NaN             NaN  NaN     NaN   
8329           NaN           NaN      NaN             NaN  NaN     NaN   
8330           NaN           NaN      NaN             NaN  NaN     NaN   
8331           NaN           NaN      NaN             NaN  NaN     NaN   

     admin_level boundary  
0            NaN      NaN  
1            NaN      NaN  
2            NaN      NaN  
3            NaN      NaN  
4            NaN      NaN  
...          ...      ...  
8327         NaN      NaN  
8328         NaN      NaN  
8329         NaN      NaN  
8330         NaN      NaN  
8331         NaN      NaN  

[8332 rows x 108 columns]

In [94]:
tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8332 entries, 0 to 8331
Columns: 108 entries, highway to boundary
dtypes: object(108)
memory usage: 6.9+ MB


In [95]:
extract_tags = [
    "name"  
]

gdf_streets = gdf_streets.join(tags[extract_tags])
gdf_streets

type         id                                               tags  \
0     way    4414959  {'highway': 'residential', 'lanes': '2', 'lit'...   
1     way    4414960  {'highway': 'residential', 'lanes': '2', 'lit'...   
2     way    4414988  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
3     way    4456239  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
4     way    4460441  {'destination': 'Россия;Москва', 'destination:...   
...   ...        ...                                                ...   
8327  way  982201974                             {'highway': 'service'}   
8328  way  982201975                             {'highway': 'service'}   
8329  way  982201976        {'access': 'private', 'highway': 'service'}   
8330  way  982309906  {'highway': 'service', 'psv': 'yes', 'surface'...   
8331  way  982465716  {'foot': 'no', 'highway': 'secondary', 'lanes'...   

                                               geometry  \
0     LINESTRING (3386560.825 8339904.673, 3386543.1...   
1     LINESTRING (3386296.074 8339493.502, 3386326.1...   
2     LINESTRING (3385061.151 8336249.836, 3385080.0...   
3     LINESTRING (3383520.646 8339537.624, 3383522.2...   
4     LINESTRING (3380805.073 8358710.316, 3380674.5...   
...                                                 ...   
8327  LINESTRING (3387389.855 8336056.523, 3387395.9...   
8328  LINESTRING (3399572.860 8344030.830, 3399577.6...   
8329  LINESTRING (3399513.438 8343712.428, 3399614.3...   
8330  LINESTRING (3386841.039 8339837.185, 3386808.8...   
8331  LINESTRING (3383406.499 8342087.601, 3383409.9...   

                           name  
0      бульвар Алексея Толстого  
1             Ахматовская улица  
2                 Садовая улица  
3     площадь Египетские Ворота  
4                           NaN  
...                         ...  
8327                        NaN  
8328                        NaN  
8329                        NaN  
8330                        NaN  
8331        Петербургское шоссе  

[8332 rows x 5 columns]

In [96]:
gdf_streets.dropna(
    how="all",  
    subset=["name"],  
    inplace=True
)
gdf_streets

type         id                                               tags  \
0     way    4414959  {'highway': 'residential', 'lanes': '2', 'lit'...   
1     way    4414960  {'highway': 'residential', 'lanes': '2', 'lit'...   
2     way    4414988  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
3     way    4456239  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
6     way    8146528  {'highway': 'trunk', 'int_ref': 'E 105;AH8', '...   
...   ...        ...                                                ...   
8302  way  975186069  {'foot': 'no', 'highway': 'secondary', 'lanes'...   
8303  way  976477993  {'addr:suburb': 'Новосёлки', 'highway': 'resid...   
8307  way  977815421  {'highway': 'residential', 'lanes': '2', 'lit'...   
8308  way  978125029  {'highway': 'service', 'name': 'Прудовая дорог...   
8331  way  982465716  {'foot': 'no', 'highway': 'secondary', 'lanes'...   

                                               geometry  \
0     LINESTRING (3386560.825 8339904.673, 3386543.1...   
1     LINESTRING (3386296.074 8339493.502, 3386326.1...   
2     LINESTRING (3385061.151 8336249.836, 3385080.0...   
3     LINESTRING (3383520.646 8339537.624, 3383522.2...   
6     LINESTRING (3384171.263 8355479.155, 3383984.0...   
...                                                 ...   
8302  LINESTRING (3383446.841 8341298.194, 3383434.5...   
8303  LINESTRING (3385384.312 8333467.983, 3385339.4...   
8307  LINESTRING (3380253.285 8345329.833, 3380238.2...   
8308  LINESTRING (3386163.181 8336220.104, 3386149.6...   
8331  LINESTRING (3383406.499 8342087.601, 3383409.9...   

                           name  
0      бульвар Алексея Толстого  
1             Ахматовская улица  
2                 Садовая улица  
3     площадь Египетские Ворота  
6              Московское шоссе  
...                         ...  
8302        Петербургское шоссе  
8303          Гражданская улица  
8307           Образцовая улица  
8308            Прудовая дорога  
8331        Петербургское шоссе  

[1807 rows x 5 columns]

In [97]:

gdf_streets.duplicated(subset=["id", "geometry", "name"]).value_counts()

False    1807
dtype: int64

In [98]:
gdf_streets["name"].is_unique

False

In [99]:
gdf_streets = gdf_streets.dissolve(by="name")  
gdf_streets

geometry  \
name                                                                          
1-й Бадаевский проезд     MULTILINESTRING ((3384252.237 8357215.464, 338...   
1-й проезд                MULTILINESTRING ((3377194.348 8340527.998, 337...   
1-я Краснофлотская улица  LINESTRING (3391605.791 8329942.456, 3391541.3...   
1-я Новая улица           LINESTRING (3382573.573 8331290.424, 3382409.7...   
1-я Советская улица       MULTILINESTRING ((3387364.508 8329267.010, 338...   
...                                                                     ...   
улица Центральная Дорога  LINESTRING (3390649.423 8321680.047, 3390800.4...   
улица Чернышевского       LINESTRING (3390526.449 8327864.622, 3390542.8...   
улица Чистякова           LINESTRING (3387046.490 8336364.198, 3387063.3...   
улица Южная Дорога        LINESTRING (3391010.321 8321139.923, 3391109.9...   
шоссе Подбельского        MULTILINESTRING ((3387663.378 8338044.794, 338...   

                         type         id  \
name                                       
1-й Бадаевский проезд     way   46102592   
1-й проезд                way   82226373   
1-я Краснофлотская улица  way  172359215   
1-я Новая улица           way  147948621   
1-я Советская улица       way   31400244   
...                       ...        ...   
улица Центральная Дорога  way  191104803   
улица Чернышевского       way  105466951   
улица Чистякова           way  193547636   
улица Южная Дорога        way  191104828   
шоссе Подбельского        way   45591891   

                                                                       tags  
name                                                                         
1-й Бадаевский проезд     {'highway': 'unclassified', 'lanes': '1', 'lit...  
1-й проезд                {'highway': 'residential', 'name': '1-й проезд...  
1-я Краснофлотская улица  {'highway': 'residential', 'name': '1-я Красно...  
1-я Новая улица           {'highway': 'service', 'name': '1-я Новая улиц...  
1-я Советская улица       {'highway': 'residential', 'lanes': '1', 'name...  
...                                                                     ...  
улица Центральная Дорога  {'highway': 'residential', 'name': 'улица Цент...  
улица Чернышевского       {'highway': 'residential', 'name': 'улица Черн...  
улица Чистякова           {'highway': 'tertiary', 'lanes': '2', 'lit': '...  
улица Южная Дорога        {'highway': 'residential', 'name': 'улица Южна...  
шоссе Подбельского        {'highway': 'unclassified', 'lanes': '2', 'lit...  

[478 rows x 4 columns]

In [100]:
gdf_streets.index.rename("Название улицы", inplace=True)
gdf_streets.head()

geometry  \
Название улицы                                                                
1-й Бадаевский проезд     MULTILINESTRING ((3384252.237 8357215.464, 338...   
1-й проезд                MULTILINESTRING ((3377194.348 8340527.998, 337...   
1-я Краснофлотская улица  LINESTRING (3391605.791 8329942.456, 3391541.3...   
1-я Новая улица           LINESTRING (3382573.573 8331290.424, 3382409.7...   
1-я Советская улица       MULTILINESTRING ((3387364.508 8329267.010, 338...   

                         type         id  \
Название улицы                             
1-й Бадаевский проезд     way   46102592   
1-й проезд                way   82226373   
1-я Краснофлотская улица  way  172359215   
1-я Новая улица           way  147948621   
1-я Советская улица       way   31400244   

                                                                       tags  
Название улицы                                                               
1-й Бадаевский проезд     {'highway': 'unclassified', 'lanes': '1', 'lit...  
1-й проезд                {'highway': 'residential', 'name': '1-й проезд...  
1-я Краснофлотская улица  {'highway': 'residential', 'name': '1-я Красно...  
1-я Новая улица           {'highway': 'service', 'name': '1-я Новая улиц...  
1-я Советская улица       {'highway': 'residential', 'lanes': '1', 'name...

In [101]:

drop_columns = [
    "id", 
    "type",
    "tags"
]
gdf_streets.drop(columns=drop_columns, inplace=True)
gdf_streets.head()

geometry
Название улицы                                                             
1-й Бадаевский проезд     MULTILINESTRING ((3384252.237 8357215.464, 338...
1-й проезд                MULTILINESTRING ((3377194.348 8340527.998, 337...
1-я Краснофлотская улица  LINESTRING (3391605.791 8329942.456, 3391541.3...
1-я Новая улица           LINESTRING (3382573.573 8331290.424, 3382409.7...
1-я Советская улица       MULTILINESTRING ((3387364.508 8329267.010, 338...

In [102]:
gdf_kgiop_objects.to_crs(4236, inplace = True)

In [103]:
gdf_kgiop_objects['lon'] = gdf_kgiop_objects.geometry.x
gdf_kgiop_objects['lat'] = gdf_kgiop_objects.geometry.y

In [104]:
gdf_kgiop_objects.to_file("kgiop2.geojson", driver = "GeoJSON")

In [105]:
gdf_streets.explore(tiles=TILES)

## Объекты культурного наследия

### Загрузка объектов культурного наследия

In [106]:
gdf_kgiop_objects = gpd.read_file(KGIOP_FILE_URL, mask=territory)
gdf_kgiop_objects

id                                      ensemble_name  \
0    6857                                Парк Екатерининский   
1    6955          Дома Царскосельского Дворцового правления   
2    6374                                                  —   
3    6375                                                  —   
4    6376                                                  —   
..    ...                                                ...   
916  9548  Казармы и службы Лейб-гвардии Сводно-Казачьего...   
917  9549  Казармы и службы Лейб-гвардии Сводно-Казачьего...   
918  9550  Казармы и службы Лейб-гвардии Сводно-Казачьего...   
919  9551  Казармы и службы Лейб-гвардии Сводно-Казачьего...   
920  9552  Казармы и службы Лейб-гвардии Сводно-Казачьего...   

                               object_name occurrence_time  \
0                                     Мост            1825   
1                                  Кузница            1808   
2                        Конюшни дворцовые       1786-1792   
3    Жилой флигель полицейского управления          1830-е   
4                           Дом Пытниковой      19 в. кон.   
..                                     ...             ...   
916                          Здание службы            1912   
917                          Здание службы            1912   
918                          Здание службы            1912   
919                     Офицерский флигель            1910   
920                                Казарма            1913   

                                       object_location  \
0                        г. Пушкин, через Рыбный канал   
1                            г. Пушкин, Малая ул., 14а   
2    г. Павловск, Красных Зорь ул., 2; Революции ул...   
3                     г. Павловск, Красных Зорь ул., 3   
4                     г. Павловск, Красных Зорь ул., 5   
..                                                 ...   
916               г. Павловск, Обороны ул., 1б, лит. Ш   
917               г. Павловск, Обороны ул., 1б, лит. Ц   
918               г. Павловск, Обороны ул., 1б, лит. Э   
919               г. Павловск, Обороны ул., 1б, лит. Ю   
920              г. Павловск, Обороны ул., 1б, лит. АБ   

                                   historical_category  \
0    объект культурного наследия федерального значения   
1               выявленный объект культурного наследия   
2    объект культурного наследия федерального значения   
3               выявленный объект культурного наследия   
4               выявленный объект культурного наследия   
..                                                 ...   
916  объект культурного наследия регионального знач...   
917  объект культурного наследия регионального знач...   
918  объект культурного наследия регионального знач...   
919  объект культурного наследия регионального знач...   
920  объект культурного наследия регионального знач...   

                                         normative_act object_type  \
0    Постановление Правительства РФ № 527 от 10.07....    Памятник   
1         Приказ председателя КГИОП № 15 от 20.02.2001    Памятник   
2    Постановление Правительства РФ № 527 от 10.07....    Памятник   
3         Приказ председателя КГИОП № 15 от 20.02.2001    Памятник   
4         Приказ председателя КГИОП № 15 от 20.02.2001    Памятник   
..                                                 ...         ...   
916           Распоряжение КГИОП от 18.10.2018 № 423-р    Памятник   
917           Распоряжение КГИОП от 18.10.2018 № 423-р    Памятник   
918           Распоряжение КГИОП от 18.10.2018 № 423-р    Памятник   
919           Распоряжение КГИОП от 18.10.2018 № 423-р    Памятник   
920           Распоряжение КГИОП от 18.10.2018 № 423-р    Памятник   

                            geometry  
0    POINT (3384134.895 8336228.624)  
1    POINT (3384169.850 8337729.962)  
2    POINT (3388966.718 8330136.117)  
3    POINT (3388870.204 8330254.332)  
4    POINT (3388846.938 8330155.525)  
..                      

In [107]:


drop_columns = [
    "id",
    "ensemble_name",
    "occurrence_time",
    "historical_category",
    "normative_act",
]

gdf_kgiop_objects.drop(columns=drop_columns, inplace=True)
gdf_kgiop_objects.head()

object_name  \
0                                   Мост   
1                                Кузница   
2                      Конюшни дворцовые   
3  Жилой флигель полицейского управления   
4                         Дом Пытниковой   

                                     object_location object_type  \
0                      г. Пушкин, через Рыбный канал    Памятник   
1                          г. Пушкин, Малая ул., 14а    Памятник   
2  г. Павловск, Красных Зорь ул., 2; Революции ул...    Памятник   
3                   г. Павловск, Красных Зорь ул., 3    Памятник   
4                   г. Павловск, Красных Зорь ул., 5    Памятник   

                          geometry  
0  POINT (3384134.895 8336228.624)  
1  POINT (3384169.850 8337729.962)  
2  POINT (3388966.718 8330136.117)  
3  POINT (3388870.204 8330254.332)  
4  POINT (3388846.938 8330155.525)

In [108]:


gdf_kgiop_objects.rename(columns = {"object_name" : "Название объекта", "object_location" : "Расположение объекта", "object_type" : "Вид объекта"}, inplace=True)
gdf_kgiop_objects.head()

Название объекта  \
0                                   Мост   
1                                Кузница   
2                      Конюшни дворцовые   
3  Жилой флигель полицейского управления   
4                         Дом Пытниковой   

                                Расположение объекта Вид объекта  \
0                      г. Пушкин, через Рыбный канал    Памятник   
1                          г. Пушкин, Малая ул., 14а    Памятник   
2  г. Павловск, Красных Зорь ул., 2; Революции ул...    Памятник   
3                   г. Павловск, Красных Зорь ул., 3    Памятник   
4                   г. Павловск, Красных Зорь ул., 5    Памятник   

                          geometry  
0  POINT (3384134.895 8336228.624)  
1  POINT (3384169.850 8337729.962)  
2  POINT (3388966.718 8330136.117)  
3  POINT (3388870.204 8330254.332)  
4  POINT (3388846.938 8330155.525)

In [109]:
gdf_kgiop_objects.explore(tiles=TILES)

In [110]:
STREET_BUFFER = 200

def get_contains_kgiop_objects(street) -> int:
    """ Функция для подсчета количества объектов культурного наследия попадающих в буффер улицы"""
    return sum(gdf_kgiop_objects["geometry"].within(street.buffer(STREET_BUFFER)))


gdf_streets["contains_kgiop_objects"] = gdf_streets["geometry"].apply(get_contains_kgiop_objects)
gdf_streets.nlargest(5, "contains_kgiop_objects")

geometry  \
Название улицы                                                        
Садовая улица     MULTILINESTRING ((3385061.151 8336249.836, 338...   
Парковая улица    MULTILINESTRING ((3388077.320 8332544.145, 338...   
Дворцовая улица   MULTILINESTRING ((3383665.595 8337133.541, 338...   
Малая улица       MULTILINESTRING ((3383667.576 8337922.392, 338...   
Московская улица  MULTILINESTRING ((3383836.448 8338218.046, 338...   

                  contains_kgiop_objects  
Название улицы                            
Садовая улица                        101  
Парковая улица                        53  
Дворцовая улица                       44  
Малая улица                           43  
Московская улица                      41

In [111]:
gdf_streets.explore("contains_kgiop_objects", tiles=TILES)

In [112]:
gdf_streets["density"] = gdf_streets["contains_kgiop_objects"] / gdf_streets.length

select_columns = []
gdf_streets.nlargest(5, "density")

geometry  \
Название улицы                                                          
Воинская дорога     LINESTRING (3382372.441 8331806.914, 3382418.1...   
Павловский мост     LINESTRING (3385080.065 8336238.005, 3385153.2...   
Руинный мост        LINESTRING (3382887.716 8338539.831, 3382912.7...   
Колокольная улица   LINESTRING (3382751.740 8331897.580, 3382704.6...   
дорога Красовского  MULTILINESTRING ((3382684.291 8331633.504, 338...   

                    contains_kgiop_objects   density  
Название улицы                                        
Воинская дорога                         16  0.119468  
Павловский мост                          9  0.103016  
Руинный мост                             4  0.083756  
Колокольная улица                        7  0.061341  
дорога Красовского                      15  0.051237

In [113]:
gdf_streets.rename(columns = {"contains_kgiop_objects" : "Количество объектов", "density" : "Плотность"}, inplace=True)
gdf_streets.head()

geometry  \
Название улицы                                                                
1-й Бадаевский проезд     MULTILINESTRING ((3384252.237 8357215.464, 338...   
1-й проезд                MULTILINESTRING ((3377194.348 8340527.998, 337...   
1-я Краснофлотская улица  LINESTRING (3391605.791 8329942.456, 3391541.3...   
1-я Новая улица           LINESTRING (3382573.573 8331290.424, 3382409.7...   
1-я Советская улица       MULTILINESTRING ((3387364.508 8329267.010, 338...   

                          Количество объектов  Плотность  
Название улицы                                            
1-й Бадаевский проезд                       0        0.0  
1-й проезд                                  0        0.0  
1-я Краснофлотская улица                    0        0.0  
1-я Новая улица                             0        0.0  
1-я Советская улица                         0        0.0

In [114]:
m = gdf_streets.explore("Количество объектов", tiles=TILES)
m = territory.explore(m=m, style_kwds={"fill": False, "weight": 3})

m

In [115]:
gdf_streets.to_file('streets_with_contains_kgiop.geojson', driver = 'GeoJSON' )

In [116]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Статистическая обработка данных



In [117]:
len(gdf_kgiop_objects) / gdf_streets.length.sum() 

0.0008194062502730812

In [118]:
gdf_streets["Количество объектов"].describe()

count    478.000000
mean       2.698745
std        8.366678
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max      101.000000
Name: Количество объектов, dtype: float64

In [119]:
gdf_streets["Количество объектов"].sort_values(ascending=False).plot.bar()

## Заключение

Пушкинский район богат на объекты культурного наследия, всего их в районе 478. 

Средняя плотность ОКН на район составляет 0.0008. 

Анализируя график, можно выделить топ 5 улиц по наличию ОКН:
- **Садовая (101),** 
- Парковая (53), 
- Дворцовая (44),
- Малая (43),
- Московская (41). 
